File: repos/koi-web/src/lib/repositories/mood_repository.ts
import { db } from '../db';
import type { MoodEntry } from '../../models/index';

export class MoodRepository {
  async save(entry: MoodEntry): Promise<void> {
    await db.moodEntries.put(entry);
  }

  async getById(id: string): Promise<MoodEntry | undefined> {
    return db.moodEntries.get(id);
  }

  async getByUser(userId: string): Promise<MoodEntry[]> {
    return db.moodEntries
      .where('userId')
      .equals(userId)
      .sortBy('timestamp');
  }

  async getByUserInRange(userId: string, from: Date, to: Date): Promise<MoodEntry[]> {
    return db.moodEntries
      .where('[userId+timestamp]')
      .between([userId, from], [userId, to], true, true)
      .toArray()
      .catch(() =>
        db.moodEntries
          .where('userId')
          .equals(userId)
          .filter((e) => e.timestamp >= from && e.timestamp <= to)
          .toArray(),
      );
  }

  async getUnsynced(): Promise<MoodEntry[]> {
    return db.moodEntries.where('synced').equals(0).toArray();
  }

  async markSynced(id: string): Promise<void> {
    await db.moodEntries.update(id, { synced: true });
  }

  async delete(id: string): Promise<void> {
    await db.moodEntries.delete(id);
  }
}

export const moodRepository = new MoodRepository();

File: repos/koi-backend/flutter_lib/lib/storage/repositories/mood_repository.dart
import 'package:hive/hive.dart';
import '../hive_manager.dart';
import '../../models/hive_mood_entry_model.dart';

class MoodRepository {
  Box<HiveMoodEntry> get _box => HiveManager.moodEntries;

  Future<void> saveEntry(HiveMoodEntry entry) async =>
      _box.put(entry.id, entry);

  HiveMoodEntry? getEntry(String id) => _box.get(id);

  List<HiveMoodEntry> getByUser(String userId) {
    return _box.values
        .where((e) => e.userId == userId)
        .toList()
      ..sort((a, b) => a.timestamp.compareTo(b.timestamp));
  }

  List<HiveMoodEntry> getByUserInRange(
      String userId, DateTime from, DateTime to) {
    return _box.values
        .where((e) =>
            e.userId == userId &&
            !e.timestamp.isBefore(from) &&
            !e.timestamp.isAfter(to))
        .toList()
      ..sort((a, b) => a.timestamp.compareTo(b.timestamp));
  }

  List<HiveMoodEntry> getUnsynced() =>
      _box.values.where((e) => !e.synced).toList();

  Future<void> markSynced(String id) async {
    final entry = _box.get(id);
    if (entry == null) return;
    entry.synced = true;
    await entry.save();
  }

  Future<void> deleteEntry(String id) async => _box.delete(id);
}
